bpf/bpf_lxc.c

```diff
@@ -544,6 +550,17 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 			*dstID = info->sec_label;
 			tunnel_endpoint = info->tunnel_endpoint;
 			encrypt_key = get_min_encrypt_key(info->key);
+#ifdef ENABLE_WIREGUARD
+			/* If we detect that the dst is a remote endpoint, we
+			 * need to mark the packet. The ip rule which matches
+			 * on the MARK_MAGIC_ENCRYPT mark will steer the packet
+			 * to the Wireguard tunnel.
+			 */
+			if (info->tunnel_endpoint != 0 &&
+			    info->sec_label != HOST_ID &&
+			    info->sec_label != REMOTE_NODE_ID)
+				ctx->mark |= MARK_MAGIC_ENCRYPT;
+#endif /* ENABLE_WIREGUARD */
```

Comment:
I'm a bit surprised to see that the mark configuration is occurring in different places from ipsec. Is there something drastically different between the two encryption models that explains this?

Reply:
> Is there something drastically different between the two encryption models that explains this?

Yes, they differ quite a lot. In IPSec we need to return a packet to the stack to encrypt/decrypt it, and route it again to the same device. While in Wireguard it's enough just to redirect the packet to the Wireguard tunnel device (`cilium_wg0`) to encrypt it (decryption happens automatically when the `cilium_wg0` receives an encrypted packet).

Reply:
Coming back to your question about the different places, I _think_ the mark could be set a bit later than it's now. Unfortunately, it cannot be set in the same place where the IPSec handling happens, as then the BPF host routing would make the packet to bypass the encryption (I haven't validated whether the BPF HR works with Wireguard though).

Reply:
Yeah I guess the way I was thinking about this is that regardless of whether we need to forward this through the xfrm stack subsystem or route to the destination, we are passing the packets up to the stack to be redirected to the destination device and from that perspective, the main thing we need to do is set the mark before we hit that code and hopefully that could be done at one consistent "encryption" stage of the BPF code rather than having different behaviour/ordering depending on the encryption implementation. (This may not be 100% accurate particularly if we start to get into BPF host routing..).

The most obvious implication right now is just the L7 handling which I already mentioned below, but in general if the encryption is handled at different points of this program, then the other logic that surrounds the encrypt logic could either run or not run depending on the mode, and that could be a source of bugs. The fewer differences there are between the BPF implementations of the two encryption approaches, the better IMO.

Reply:
Makes sense. For now I've moved setting the mark to the same place as we do for IPSec. We might need to move it higher once we start supporting BPF HR.